### DISCUSSION TITLE
Forms setters ans constraints execution ordering

### DISCUSSION THREAD

**Discussion by CODEheures** [symfony/symfony#46707] [Q&A] [answered]
Why form constraints (based on form->getData) are executed after form setters

This is a problem if setter must set a non nullable typed property
Form constraint NotBlank should prevent setter execution in this case

And other observation: Because I have to manage null case in form setter for non nullable properties, constraint in form has not the same effect as Assert on entity because constraints are testing form data and Assert are testing Entity property that are not set for null value

Finally, I will conclude forms getter/setter/constraints are recommanded for non nullable typed properties

**Comment by ghost** [comment, marked as answer]
You get it all wrong.
Setters must be nullable because you don't know whether a field will come within a request (i.e. will be missing or empty - read more what will happen then here: https://symfony.com/doc/current/reference/forms/types/form.html#empty-data). This is why you use constraints - to validate if required field is not blank. Data passed to form when submitting is first mapped to the model (with the help from PropertyAccess Component) and then, as the last step (in POST_SUBMIT event), the model is validated (with the help from Validator Component). 
With that said, people often create anemic form models and then, when validated, pass validated data to the entity that represents business model where e.g. some properties cannot be null.

**Reply by CODEheures** [reply]
Thanks for response. 

Ok, for nullable setter and not blank constraint on form in case of PHP7 not nullable typed property to prevent fails when calling entity setter.

But, for other constraints (ex: min length), you duplicate there too between Form Constraints and Entity Assets? (Or you only use one, which one? Are there any best practices?)

**Reply by ghost** [reply]
I would say that your entity should always be in a valid state, i.e. you don't put asserts on the entities because at the moment of passing data from form model to the entity you know the data is valid. 
Some people add one more level of safety and use domain validations inside entities e.g. inside setters body or take advantage of Value Objects which are always valid by definition and use them instead of primitive types as parameter types in domain layer.

**Reply by CODEheures** [reply]
Ok thanks a lot for response :-)
